core/src/sumeragi/main_loop.rs

```diff
@@ -669,9 +672,15 @@ fn reset_state(
     voting_block: &mut Option<VotingBlock>,
     voting_signatures: &mut Vec<SignatureOf<SignedBlock>>,
     round_start_time: &mut Instant,
+    last_view_change_time: &mut Instant,
     view_change_time: &mut Duration,
+    node_wants_block: &mut bool,
 ) {
-    let mut was_commit_or_view_change = current_latest_block_height != *old_latest_block_height;
+    let mut was_commit_or_view_change = false;
+    if current_latest_block_height != *old_latest_block_height {
+        *round_start_time = Instant::now();
```

Comment:
Is this change relevant for this PR?

I suspect this change will bring about issues that would require the logic to become more convoluted. I will try to recall the problems if you think this change is relevant.

Reply:
is this because you want to prevent view changes when there are no transactions?

Reply:
Yea it is in order to differentiate between when consensus is working and when it is not. You would be correct that it doesn't do that much now but for full adherence to the whitepaper this is an important distinction. For instance, now we shuffle all the peers for each view change or block commit. In paper you only shuffle within the A set when things are working, and only when consensus fails do you bring in observing peers. Basically when things are going well observing peers are never involved in block production.

Reply:
https://github.com/hyperledger/iroha/blob/iroha2-dev/docs/source/iroha_2_whitepaper.md#28-consensus

Reply:
> For each round (e.g., block), the previous round's (block's) hash is used to determine an ordering over the set of nodes, such that there is a deterministic and canonical order for each block. In this ordering, the first nodes are grouped into a set called set . Under normal (non-faulty) conditions, consensus for a block is performed by set . The remaining nodes are grouped into a set called set . Under normal conditions, set acts as a passive set of validators to view and receive committed blocks, but otherwise they do not participate in consensus.

It's seems like we actually change set A on every new block.

Reply:
> For instance, now we shuffle all the peers for each view change or block commit. In paper you only shuffle within the A set when things are working, and only when consensus fails do you bring in observing peers. Basically when things are going well observing peers are never involved in block production

I would agree that we shouldn't involve observing peers when things go well. This way we're not wasting time on view changes to find A set that contains no faulty peers. I believe this is important if we want to maintain high throughput in an environment with a lot of faulty peers (`~f`).
However, in the case when there are no transactions, forgetting which nodes comprised a valid A set, and consequently wasting a few `sumeragi` cycles to find it again, is of little relevance.
Also, if we don't do a view change when there are no transactions we're going to have to bring back that mechanism for forwarding a transaction to a leader to make sure leader isn't faulty

I strive for simplicity. If this isn't convincing I will go about to find a loophole in the logic you implemented because as far as I remember there was one

Reply:
just add the following comment above (or you can find a more appropriate place):

```suggestion
        // Round is only restarted on a block commit, so that in the case of 
        // a view change a new block is immediately created by the leader
        *round_start_time = Instant::now();
```

Reply:
will do.